lstchain/reco/utils.py

```diff
@@ -611,3 +612,26 @@ def clip_alt(alt):
     to keep astropy happy
     """
     return np.clip(alt, -90.*u.deg, 90.*u.deg)
+
+def get_effective_time(events):
+    """
+    Calculate the effective observation time of a set 
+    of real data events.
+    Parameters
+    ----------
+    events: pandas DataFrame
+
+    Returns
+    -------
+    t_eff: float
+    t_elapsed: float
+    """
+    
+    deltat = np.diff(events['dragon_time'])
+    deltat = deltat[(deltat > 0) & (deltat < 0.002)]
+    rate=1/np.mean(deltat)
+    dead_time = np.amin(deltat)
+    t_elapsed = len(events)/rate * u.s
+    t_eff = t_elapsed/(1+rate*dead_time)
```

Comment:
This is not really correct if applied to DL2 events, unless all triggered events are present in the file, which I think we cannot assume to be the case. Are interleaved pedestal and flatfield events still there? (event if they are now, they might eventually be filtered). Also: events for which no or too few pixels survived cleaning may not be there.

Reply:
> Are interleaved pedestal and flatfield events still there?

Yes

> events for which no or too few pixels survived cleaning may not be there.

Not all the events are present in the DL2 files. There are less than the initial 53k that are present in the DL1 files.

Reply:
I think then we should run this immediately and store it in the metadata of the dl1 files, right?

Reply:
We can solve this if we save in the DL1 file a parameter "delta_t" which is the time difference between each event and _the previously triggered_ one.

Reply:
Yes, storing delta_t for all events will also just work.

Reply:
instead of storing an additional parameter for all events, can we apply this to DL1 data better so we do not increase the data volume as @maxnoe was proposing above? Is it foreseen that this `delta_t` is needed for anything else apart than calculating the effective time?

Reply:
Calculating livetime in DL1 is not flexible enough. You won't be able to apply e.g. zenith cuts later, except in the chunks for which livetime is precalculated

Reply:
> instead of storing an additional parameter for all events, can we apply this to DL1 data better so we do not increase the data volume as @maxnoe was proposing above? Is it foreseen that this `delta_t` is needed for anything else apart than calculating the effective time?

Not really "needed" but it is a useful diagnostic tool in some cases. If something weird appears in the data you can check if it is related with too close events (e.g. related to the DRS4 corrections)

Reply:
> Well we need to implement the exponential fit to get the external rate (dead-time-free) instead of the observed event rate.

The mean is the maximum-likelihood estimator for the exponential and the min is the likelihood estimator for the deat time. No need to perform some kind of explicit fit, I think.

Reply:
> @rlopezcoto : Calculating livetime in DL1 is not flexible enough. You won't be able to apply e.g. zenith cuts later, except in the chunks for which livetime is precalculated

You want to apply zenith cuts *within* a run?

Reply:
> > @rlopezcoto : Calculating livetime in DL1 is not flexible enough. You won't be able to apply e.g. zenith cuts later, except in the chunks for which livetime is precalculated
> 
> You want to apply zenith cuts _within_ a run?

We don't have a predetermined run duration yet, though typically they are short (20').  
But if you want to be able to apply a precise zenith cut to a sample, you'll be forced to cut inside runs, or apply a rough cuts on entire runs. I think the former approach is better, and then you'll want to have the effective time.

Reply:
> > Well we need to implement the exponential fit to get the external rate (dead-time-free) instead of the observed event rate.
> 
> The mean is the maximum-likelihood estimator for the exponential and the min is the likelihood estimator for the dead time. No need to perform some kind of explicit fit, I think.

You are right, it is just slightly modified in the case dead time is present:

<delta_t> = d + 1 / lambda

Reply:
Now taking into account interleaved events at a regular rate lambda_interleaved (@FrancaCassol: are flat-field & pedestal interleaved events exactly interleaved? I mean alternating ff-P-ff-P-.... with same interval between each ff andP):

- Store **delta_t** (time from previous triggered event) in DL1
- Find dead time per event: **d**  (from minimum observed delta_t)
- Make histogram of **delta_t**   **for cosmic events only**  (will anyway be truncated at the time interval between interleaved events)
- fit histogram to an exponential to obtain the true cosmic event rate: **lambda**
- calculate elapsed time (just add up the times between timestamps of **DL2 events**, excluding large values which indicate observation gaps): **t_elapsed**
- Obtain effective time as  **t_eff = t_elapsed / (1+lambda d + lambda_interleaved d)**

Reply:
Ok! After reading all the discussion, seems like there's some work to do before merging this function. We must then work on the storage of delta_t in DL1 files, shall we do that in a separate PR?

Reply:
Assuming lambda=10kHz, lambda_interleaved=200Hz, d = 7e-6 s  The error from neglecting the interleaved term above is just  0.13%  in t_eff. We can also ignore it for now and just write a note about it in the code.

Reply:
> Ok! After reading all the discussion, seems like there's some work to do before merging this function. We must then work on the storage of delta_t in DL1 files, shall we do that in a separate PR?

I don't think another PR is needed. This is precisely fot t_eff calculation, right? So why separate it?

Reply:
Yes! no problem, just asking, then let's implement it here.

Reply:
@moralejo: the problem is that now this needs to be implemented in the `r0_to_dl1` and this would better be suited for `ctapipe_io_lst`, at least for the estable long term version

Reply:
`ctapipe_io_lst` fills the ctapipe `ArrayEventContainer` which has now field for a `delta_t`, so I think the right place would be `r0_to_dl1`.

Reply:
Indeed, r0_to_dl1.

Reply:
ok for the time being while we still use r0_to_dl1, but what about when we definitely move to the stage1 tool?

Reply:
I don't know if there is anything foreseen, since for now stage-1 is more MC-oriented. @maxnoe do you know if there is any discussion on this on the ctapipe side?

Reply:
I think this has to be dealed with "the proper way" in ctapipe. There is no such info yet, but I imaging the way this is planned to be dealt with in CTA is a monitoring table where the accumulated dead time is store every couple of seconds?

How will LST provide this information? Estimating this from event distributions is probably not the best way.

Reply:
In FACT we store the dead time fraction every 5 seconds.

Reply:
The relevant timescale seems to be 10 s: https://jama.cta-observatory.org/perspective.req#/items/26761?projectId=11 , and a monitoring table seems to be the place to keep it, yes.

Reply:
> How will LST provide this information? Estimating this from event distributions is probably not the best way.

We could just assume fixed deadtime per triggered event, but I think that would be less accurate. The delta_t distribution does not have a perfectly 'sharp rise' at the expected deadtime per event.

I'd say that when we move to stage1 we should make this in a 'ctapipe compliant' way, perhaps filling the said monitoring table in the R1 event source.

Reply:
Wait, shouldn't be there some hardware component for LST that just writes this information?

Reply:
> Wait, shouldn't be there some hardware component for LST that just writes this information?

As said in the recent call: not that I know of, but with the trigger rate and the known (or empirically determined) dead time per event it should be straightforward to obtain that, e.g. in the R1 event source.

Reply:
See also #615 , I implemented a slightly simplified version of the method outlined above.